### PR TITLE
Improve clarity of errors for GEO commands when member does not exist

### PR DESCRIPTION
**Pull request by chx9**
this pr aims to fix one of the issues from Redis Triage
https://github.com/orgs/redis/projects/6?pane=issue&itemId=69775417&issue=redis%7Credis%7C13350
it has been more than 1 year since this issue brought up

### CODE REVIEW

**Comment by Fusl** [comment]
`longLatFromMember` can also return `C_ERR` if `geohashDecodeToLongLatType` returns a decoding error while the member did exist. Should we maybe differentiate those cases or just return `member %s does not exist or failed to decode`?

**Comment by madolson** [comment]
> longLatFromMember can also return C_ERR if geohashDecodeToLongLatType returns a decoding error while the member did exist. Should we maybe differentiate those cases or just return member %s does not exist or failed to decode?

Ideally yes. It would be nice to differentiate the member be missing from some type of corruption.

**Comment by chx9** [comment]
> > longLatFromMember can also return C_ERR if geohashDecodeToLongLatType returns a decoding error while the member did exist. Should we maybe differentiate those cases or just return member %s does not exist or failed to decode?
> 
> Ideally yes. It would be nice to differentiate the member be missing from some type of corruption.

agree, i added error infomation for each case.

File: src/geo.c [hunk 1]
```diff
@@ -112,11 +112,17 @@ int extractLongLatOrReply(client *c, robj **argv, double *xy) {
 /* Input Argument Helper */
 /* Decode lat/long from a zset member's score.
  * Returns C_OK on successful decoding, otherwise C_ERR is returned. */
-int longLatFromMember(robj *zobj, robj *member, double *xy) {
+int longLatFromMember(robj *zobj, robj *member, double *xy, sds* err) {
```
Comment (madolson) [inline src/geo.c]:
```suggestion
int longLatFromMemberOrReply(client *c, robj *zobj, robj *member, double *xy) {
```
A more common pattern we use is "validate" or reply. This makes it so we don't have to pass a buncch of stack allocated objects back and forth. If we pass in the client object, we can do all the logic and then also put the replies on the client object here. The replying code becomes centralized as well.

**Review by madolson — COMMENTED** [review]
Functionality looks good, I have a high level comment for trying to make it a little bit simpler. Let me know if it makes sense.

File: src/geo.c [hunk 2]
```diff
@@ -112,11 +112,17 @@ int extractLongLatOrReply(client *c, robj **argv, double *xy) {
...
     double score = 0;
 
-    if (zsetScore(zobj, member->ptr, &score) == C_ERR) return C_ERR;
-    if (!decodeGeohash(score, xy)) return C_ERR;
+    if (zsetScore(zobj, member->ptr, &score) == C_ERR){
+        *err = sdsnew("member des not exist");
```
Comment (zuiderkwast) [inline src/geo.c]: Spelling: des -> does

File: src/geo.c (hunk 1 above)
Comment (chx9) [reply src/geo.c]: yeah, its cleaner, i fixed it

File: src/geo.c (hunk 2 above)
Comment (chx9) [reply src/geo.c]: thanks, fixed

File: tests/unit/geo.tcl [hunk 3]
```diff
@@ -345,7 +351,12 @@ start_server {tags {"geo"}} {
         set ret [r GEORADIUSBYMEMBER k1 n1 5009431 m]
         assert_equal $ret {n1 n2}
     }
-
+    test {GEOSEARCH FROMMEMBER member does not exist} {
```
Comment (madolson) [inline tests/unit/geo.tcl:354]: Code looks good. If you can add a test for the negative case for the invalid geohash, (you can just do ZADD on the Sicilty object with some invalid data) I think it'll be good to go.

Comment (chx9) [reply tests/unit/geo.tcl:354]: i found out decodeGeohash could never return 0, maybe i should remove the second addReplyErrorFormat, what do you think？

Comment (madolson) [reply tests/unit/geo.tcl:354]: You're right, invalid values get clamped into the "valid" range of values and returned, even when they should probably be returned as invalid. I think we can just leave it as a defensive check.

Comment (chx9) [reply tests/unit/geo.tcl:354]: done, I left the invalid return val, but can't add tests for that.

Comment (chx9) [reply tests/unit/geo.tcl:354]: done, I left the invalid return val, but can't add tests for that.

**Review by zuiderkwast — APPROVED** [review]
LGTM

**Review by madolson — APPROVED** [review]
Thanks!
